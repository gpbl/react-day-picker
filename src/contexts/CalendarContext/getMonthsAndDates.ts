import { startOfDay, startOfMonth } from 'date-fns';

import { FormatOptions } from '../../types/FormatOptions';

import { getDates } from './utils/getDates';
import { getDayPickerMonths } from './utils/getDayPickerMonths';
import { getDisplayMonths } from './utils/getDisplayMonths';

export function getMonthsAndDates(
  firstMonth: Date,
  toDate: Date | undefined,
  options: {
    numberOfMonths: number;
    reverseMonths?: boolean;
    ISOWeek?: boolean;
    fixedWeeks?: boolean;
    locale: FormatOptions['locale'];
    weekStartsOn?: FormatOptions['weekStartsOn'];
    firstWeekContainsDate?: FormatOptions['firstWeekContainsDate'];
  }
) {
  const { numberOfMonths, reverseMonths, ISOWeek, fixedWeeks } = options;

  const firstDayOfFirstMonth = startOfMonth(startOfDay(firstMonth));

  const displayMonths = getDisplayMonths(
    firstDayOfFirstMonth,
    numberOfMonths,
    toDate
  );
  const lastMonth = displayMonths[displayMonths.length - 1];
  const dates = getDates(firstDayOfFirstMonth, lastMonth, toDate, {
    ISOWeek,
    fixedWeeks,
    locale: options?.locale,
    weekStartsOn: options?.weekStartsOn
  });
  const months = getDayPickerMonths(displayMonths, dates, {
    reverseMonths,
    ISOWeek,
    fixedWeeks,
    locale: options?.locale,
    weekStartsOn: options?.weekStartsOn,
    firstWeekContainsDate: options?.firstWeekContainsDate
  });

  return { dates, months };
}
