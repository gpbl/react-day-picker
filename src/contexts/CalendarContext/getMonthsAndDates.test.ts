import { getMonthsAndDates } from './getMonthsAndDates';

const today = new Date(2023, 9, 19);

jest.setSystemTime(today);

const firstMonth = today;
const toDate = undefined;

describe('when using the default options', () => {
  const calendar = getMonthsAndDates(firstMonth, toDate, {
    numberOfMonths: 1,
    reverseMonths: false,
    ISOWeek: false,
    fixedWeeks: false,
    locale: undefined,
    weekStartsOn: undefined,
    firstWeekContainsDate: undefined
  });
  test('should match the snapshot', () => {
    expect(calendar).toMatchSnapshot();
  });
  test('should return 1 month', () => {
    expect(calendar.months).toHaveLength(1);
  });
  test('should return 42 dates', () => {
    expect(calendar.dates).toHaveLength(35);
  });
  test('the first date should be sunday', () => {
    expect(calendar.dates[0].getDay()).toBe(0);
  });
  test('the last date should be sunday', () => {
    expect(calendar.dates[0].getDay()).toBe(0);
  });
});

describe('when using fixed weeks', () => {
  const calendar = getMonthsAndDates(today, toDate, {
    numberOfMonths: 1,
    reverseMonths: false,
    ISOWeek: false,
    fixedWeeks: true,
    locale: undefined,
    weekStartsOn: undefined,
    firstWeekContainsDate: undefined
  });
  test('should return 42 dates', () => {
    expect(calendar.dates).toHaveLength(42);
  });
  test('dates should match the snapshot', () => {
    expect(calendar.dates).toMatchSnapshot();
  });
});

describe('when using ISO week dates', () => {
  const calendar = getMonthsAndDates(today, toDate, {
    numberOfMonths: 1,
    reverseMonths: false,
    ISOWeek: true,
    fixedWeeks: false,
    locale: undefined,
    weekStartsOn: undefined,
    firstWeekContainsDate: undefined
  });

  test('should return 42 dates', () => {
    expect(calendar.dates).toHaveLength(42);
  });
  test('dates should match the snapshot', () => {
    expect(calendar).toMatchSnapshot();
  });
  test('first week number should be 39', () => {
    expect(calendar.months[0].weeks[0].weekNumber).toBe(39);
  });
  test('last day of the first week should be 1', () => {
    expect(calendar.months[0].weeks[0].days[6].date.getDate()).toBe(1);
  });
});
