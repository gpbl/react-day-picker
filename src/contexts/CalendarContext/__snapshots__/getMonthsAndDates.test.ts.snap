// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when using ISO week dates dates should match the snapshot 1`] = `
{
  "dates": [
    2023-09-25T05:00:00.000Z,
    2023-09-26T05:00:00.000Z,
    2023-09-27T05:00:00.000Z,
    2023-09-28T05:00:00.000Z,
    2023-09-29T05:00:00.000Z,
    2023-09-30T05:00:00.000Z,
    2023-10-01T05:00:00.000Z,
    2023-10-02T05:00:00.000Z,
    2023-10-03T05:00:00.000Z,
    2023-10-04T05:00:00.000Z,
    2023-10-05T05:00:00.000Z,
    2023-10-06T05:00:00.000Z,
    2023-10-07T05:00:00.000Z,
    2023-10-08T05:00:00.000Z,
    2023-10-09T05:00:00.000Z,
    2023-10-10T05:00:00.000Z,
    2023-10-11T05:00:00.000Z,
    2023-10-12T05:00:00.000Z,
    2023-10-13T05:00:00.000Z,
    2023-10-14T05:00:00.000Z,
    2023-10-15T05:00:00.000Z,
    2023-10-16T05:00:00.000Z,
    2023-10-17T05:00:00.000Z,
    2023-10-18T05:00:00.000Z,
    2023-10-19T05:00:00.000Z,
    2023-10-20T05:00:00.000Z,
    2023-10-21T05:00:00.000Z,
    2023-10-22T05:00:00.000Z,
    2023-10-23T05:00:00.000Z,
    2023-10-24T05:00:00.000Z,
    2023-10-25T05:00:00.000Z,
    2023-10-26T05:00:00.000Z,
    2023-10-27T05:00:00.000Z,
    2023-10-28T05:00:00.000Z,
    2023-10-29T05:00:00.000Z,
    2023-10-30T05:00:00.000Z,
    2023-10-31T05:00:00.000Z,
    2023-11-01T05:00:00.000Z,
    2023-11-02T05:00:00.000Z,
    2023-11-03T05:00:00.000Z,
    2023-11-04T05:00:00.000Z,
    2023-11-05T05:00:00.000Z,
  ],
  "months": [
    DayPickerMonth {
      "date": 2023-10-01T05:00:00.000Z,
      "weeks": [
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-09-25T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-09-26T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-09-27T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-09-28T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-09-29T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-09-30T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-10-01T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 39,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-02T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-03T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-04T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-05T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-06T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-07T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-08T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 40,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-09T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-10T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-11T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-12T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-13T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-14T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-15T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 41,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-16T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-17T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-18T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-19T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-20T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-21T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-22T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 42,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-23T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-24T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-25T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-26T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-27T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-28T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-29T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 43,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-30T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-31T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-11-01T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-11-02T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-11-03T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-11-04T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-11-05T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
          ],
          "weekNumber": 44,
        },
      ],
    },
  ],
}
`;

exports[`when using fixed weeks dates should match the snapshot 1`] = `
[
  2023-10-01T05:00:00.000Z,
  2023-10-02T05:00:00.000Z,
  2023-10-03T05:00:00.000Z,
  2023-10-04T05:00:00.000Z,
  2023-10-05T05:00:00.000Z,
  2023-10-06T05:00:00.000Z,
  2023-10-07T05:00:00.000Z,
  2023-10-08T05:00:00.000Z,
  2023-10-09T05:00:00.000Z,
  2023-10-10T05:00:00.000Z,
  2023-10-11T05:00:00.000Z,
  2023-10-12T05:00:00.000Z,
  2023-10-13T05:00:00.000Z,
  2023-10-14T05:00:00.000Z,
  2023-10-15T05:00:00.000Z,
  2023-10-16T05:00:00.000Z,
  2023-10-17T05:00:00.000Z,
  2023-10-18T05:00:00.000Z,
  2023-10-19T05:00:00.000Z,
  2023-10-20T05:00:00.000Z,
  2023-10-21T05:00:00.000Z,
  2023-10-22T05:00:00.000Z,
  2023-10-23T05:00:00.000Z,
  2023-10-24T05:00:00.000Z,
  2023-10-25T05:00:00.000Z,
  2023-10-26T05:00:00.000Z,
  2023-10-27T05:00:00.000Z,
  2023-10-28T05:00:00.000Z,
  2023-10-29T05:00:00.000Z,
  2023-10-30T05:00:00.000Z,
  2023-10-31T05:00:00.000Z,
  2023-11-01T05:00:00.000Z,
  2023-11-02T05:00:00.000Z,
  2023-11-03T05:00:00.000Z,
  2023-11-04T05:00:00.000Z,
  2023-11-05T05:00:00.000Z,
  2023-11-06T05:00:00.000Z,
  2023-11-07T05:00:00.000Z,
  2023-11-08T05:00:00.000Z,
  2023-11-09T05:00:00.000Z,
  2023-11-10T05:00:00.000Z,
  2023-11-11T05:00:00.000Z,
]
`;

exports[`when using the default options should match the snapshot 1`] = `
{
  "dates": [
    2023-10-01T05:00:00.000Z,
    2023-10-02T05:00:00.000Z,
    2023-10-03T05:00:00.000Z,
    2023-10-04T05:00:00.000Z,
    2023-10-05T05:00:00.000Z,
    2023-10-06T05:00:00.000Z,
    2023-10-07T05:00:00.000Z,
    2023-10-08T05:00:00.000Z,
    2023-10-09T05:00:00.000Z,
    2023-10-10T05:00:00.000Z,
    2023-10-11T05:00:00.000Z,
    2023-10-12T05:00:00.000Z,
    2023-10-13T05:00:00.000Z,
    2023-10-14T05:00:00.000Z,
    2023-10-15T05:00:00.000Z,
    2023-10-16T05:00:00.000Z,
    2023-10-17T05:00:00.000Z,
    2023-10-18T05:00:00.000Z,
    2023-10-19T05:00:00.000Z,
    2023-10-20T05:00:00.000Z,
    2023-10-21T05:00:00.000Z,
    2023-10-22T05:00:00.000Z,
    2023-10-23T05:00:00.000Z,
    2023-10-24T05:00:00.000Z,
    2023-10-25T05:00:00.000Z,
    2023-10-26T05:00:00.000Z,
    2023-10-27T05:00:00.000Z,
    2023-10-28T05:00:00.000Z,
    2023-10-29T05:00:00.000Z,
    2023-10-30T05:00:00.000Z,
    2023-10-31T05:00:00.000Z,
    2023-11-01T05:00:00.000Z,
    2023-11-02T05:00:00.000Z,
    2023-11-03T05:00:00.000Z,
    2023-11-04T05:00:00.000Z,
  ],
  "months": [
    DayPickerMonth {
      "date": 2023-10-01T05:00:00.000Z,
      "weeks": [
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-01T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-02T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-03T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-04T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-05T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-06T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-07T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 40,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-08T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-09T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-10T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-11T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-12T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-13T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-14T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 41,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-15T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-16T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-17T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-18T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-19T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-20T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-21T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 42,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-22T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-23T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-24T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-25T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-26T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-27T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-28T05:00:00.000Z,
              "displayMonth": undefined,
            },
          ],
          "weekNumber": 43,
        },
        DayPickerWeek {
          "days": [
            DayPickerDay {
              "date": 2023-10-29T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-30T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-10-31T05:00:00.000Z,
              "displayMonth": undefined,
            },
            DayPickerDay {
              "date": 2023-11-01T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-11-02T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-11-03T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
            DayPickerDay {
              "date": 2023-11-04T05:00:00.000Z,
              "displayMonth": 2023-10-01T05:00:00.000Z,
            },
          ],
          "weekNumber": 44,
        },
      ],
    },
  ],
}
`;
